### PR TITLE
Add a test to verify finding fragment id in non UTF8 document

### PR DESCRIPTION
**Pull request by rwlbuis**

### CODE REVIEW

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+<meta name=timeout content=long>
+<meta http-equiv="Content-Type" content="text/html; charset=gbk"/>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+<div></div>
+<div id="&#x586f" style="position:absolute; top:100px;"></div>
+<div style="height:200em;"></div>
+<script>
+var steps = [{
```
Comment (foolip) [inline html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html:12]: It looks like this style is copied from sibling tests which have more than one step. With just one step, I find it quite hard to follow with all the indirection.

Can you try flattening it into a simple single-page test, be something like this?
```js
setup({ single_test: true });
assert_equals(location.hash, "", "Page must be loaded with no hash");
location.hash = '%89g';
setTimeout(() => {
  assert_equals( document.scrollingElement.scrollTop, 0 );
  done();
}, 1);
```

I think that's all the test is doing, but I haven't tested my suggestion.

**Review by foolip — COMMENTED** [review]
Can you link to the part of the spec this is testing? Is it a recent change?

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html (hunk 1 above)
Comment (rwlbuis) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html]: Thanks! I tested it and it works, so I pushed that new version. Initially I thought perhaps we will add more subtests so I kept the multiple tests approach, but I guess we can introduce it later if needed.

**Comment by rwlbuis** [comment]
> Can you link to the part of the spec this is testing? Is it a recent change?

This is related to https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid:the-indicated-part-of-the-document-6. See also https://github.com/whatwg/html/issues/2902.

In general this comes from my fix in https://bugs.webkit.org/show_bug.cgi?id=83930, i.e. make scroll frag id behavior according to spec in WebKit. I think the change is a bit older but WebKit did not get around to fixing it. The fix works but we found out the test anchor-frames-gbk.html would have a different behavior (since it uses a non UTF8 document), hence this test which should replace it (anchor-frames-gbk.html is used in chromium too btw).

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+<meta http-equiv="Content-Type" content="text/html; charset=gbk"/>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+<div></div>
+<div id="&#x586f" style="position:absolute; top:100px;"></div>
+<div style="height:200em;"></div>
+<script>
+setup({ single_test: true });
+assert_equals(location.hash, "", "Page must be loaded with no hash");
```
Comment (annevk) [inline html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html:13]: Maybe also assert that document.characterSet is GBK?

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<!doctype html>
+<title>Fragment Navigation: fragment id should not be found in non UTF8 document</title>
+<meta name=timeout content=long>
+<meta http-equiv="Content-Type" content="text/html; charset=gbk"/>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<body>
+<div></div>
+<div id="&#x586f" style="position:absolute; top:100px;"></div>
+<div style="height:200em;"></div>
```
Comment (annevk) [inline html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html:10]: Don't you want 200vh or some such?

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+<script>
+setup({ single_test: true });
+assert_equals(location.hash, "", "Page must be loaded with no hash");
+location.hash = '%89g';
+setTimeout(() => {
```
Comment (annevk) [inline html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html:15]: I'm surprised this passes the lint check. Using an async_test wrapper and t.step_timeout seems slightly preferable?

**Review by annevk — COMMENTED** [review]
Thanks for writing this!

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html (hunk 2 above)
Comment (rwlbuis) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html:13]: Done.

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html (hunk 3 above)
Comment (rwlbuis) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html]: Sure, done. BTW scroll-frag-percent-encoded.html uses 200em.

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html (hunk 4 above)
Comment (rwlbuis) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html]: Not sure if a lint checker can catch this. Anyway, fixed.

Comment (annevk) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html]: As far as I know unless this directory is in lint.ignore (I guess it might be given your other comment) the literal string setTimeout will trip it.

**Review by annevk — APPROVED** [review]

File: html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html (hunk 4 above)
Comment (foolip) [reply html/browsers/browsing-the-web/scroll-to-fragid/scroll-frag-non-utf8-encoded-document.html]: It's because `lint.ignore` has `SET TIMEOUT: html/browsers/browsing-the-web/scroll-to-fragid/*`.
